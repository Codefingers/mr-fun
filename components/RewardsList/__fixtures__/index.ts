import type { Reward } from "../../../generated/graphql";

export const mockedRewards = [
  {
    id: '__TEST_REWARD_ID_1__',
    name: '__TEST_REWARD_NAME_1__',
    image: '__TEST_IMAGE_URL_1__',
    description: '__TEST_DESCRIPTION_1__',
    points: 5
  },
  {
    id: '__TEST_REWARD_ID_2__',
    name: '__TEST_REWARD_NAME_2__',
    image: '__TEST_IMAGE_URL_2__',
    description: '__TEST_DESCRIPTION_2__',
    points: 500
  },
  {
    id: '__TEST_REWARD_ID_3__',
    name: '__TEST_REWARD_NAME_3__',
    image: '__TEST_IMAGE_URL_3__',
    description: '__TEST_DESCRIPTION_3__',
    points: 7
  }
] as unknown as Array<Reward>

export const mockedSortedRewards = [
  {
    id: '__TEST_REWARD_ID_2__',
    name: '__TEST_REWARD_NAME_2__',
    image: '__TEST_IMAGE_URL_2__',
    description: '__TEST_DESCRIPTION_2__',
    points: 500
  },
  {
    id: '__TEST_REWARD_ID_3__',
    name: '__TEST_REWARD_NAME_3__',
    image: '__TEST_IMAGE_URL_3__',
    description: '__TEST_DESCRIPTION_3__',
    points: 7
  },
  {
    id: '__TEST_REWARD_ID_1__',
    name: '__TEST_REWARD_NAME_1__',
    image: '__TEST_IMAGE_URL_1__',
    description: '__TEST_DESCRIPTION_1__',
    points: 5
  }
]