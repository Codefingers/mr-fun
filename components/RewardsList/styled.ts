import * as SeparatorPrimitive from '@radix-ui/react-separator';
import styled from 'styled-components'

import { device } from '../../theme';

export const List = styled.ul`
  margin: 0 ${({ theme }) => theme.margin[2]};
  padding: 0;
  list-style: none;
  display: grid;
  grid-template-columns: repeat(1, 1fr);
  column-gap: ${({ theme }) => theme.spacing[2]};
  row-gap: ${({ theme }) => theme.spacing[2]};

  @media ${device.tablet} {
    grid-template-columns: repeat(2, 1fr);
  }

  @media ${device.laptop} {
    grid-template-columns: repeat(3, 1fr);
  }

  @media ${device.laptopL} {
    grid-template-columns: repeat(4, 1fr);
  }

  li {
    min-width: 0;
    min-height: 0;
  }

  li:last-of-type {
    margin-bottom: 0;
  }
`

export const Toolbar = styled.div`
  display: flex;
  flex-direction: column;
  margin: ${({ theme }) => theme.margin[2]};
`

export const Separator = styled(SeparatorPrimitive.Root)`
  background-color: ${({ theme }) => theme.colors.lightGrey};
  margin: ${({ theme }) => theme.margin[3]} ${({ theme }) => theme.margin[2]};

  &[data-orientation=horizontal] {
    height: 4px;
    width: auto;
  }
`