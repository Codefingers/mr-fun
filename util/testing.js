import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

const Providers = ({ children }) => {
  return (
    <ThemeProvider
      theme={{
        colors: {
          body: '#838fa6',
          heading: '#ebedf2',
          accent: '#BA99FF',
          button: '#FFF',
        },
      }}
    >
      {children}
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';

// override render method
export { customRender as render };
