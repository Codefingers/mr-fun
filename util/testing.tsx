import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../theme';

export const Providers = ({ children }) => {
  return (
    <ThemeProvider
      theme={theme}
    >
      {children}
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';

// override render method
export { customRender as render };
