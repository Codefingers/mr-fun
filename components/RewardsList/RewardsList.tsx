import RewardCard from '../../components/RewardCard';
import ToggleGroup from '../../components/ToggleGroup';

import { LIST_MUTATION_KEYS } from '../../constants';

import type { Reward } from '../../generated/graphql';

// import { useRewardsFiltering } from './hooks/useRewardsFiltering';
import { useRewardsOrdering } from './hooks/useRewardsOrdering';

import * as Styled from './styled';

interface RewardsListProps {
  rewards: Array<Reward>
}

const RewardsList = ({ rewards }: RewardsListProps) => {
  const { orderedRewards, onOrderKeyChange } = useRewardsOrdering(rewards)

  const rewardsToDisplay = orderedRewards.length ? orderedRewards : rewards

  /**
   * Opted to leave out the filtering due to already spending over the recommended time 
   * and not feeling totally happy with the route of my implementation.
   * I've left this code here and will leave the filter hook in place so you can 
   * at least get an idea of where I was planning to go with it.
   */
  // const { filteredRewards, onFilterKeyChange } = useRewardsFiltering(rewards)
  // const { orderedRewards, onOrderKeyChange } = useRewardsOrdering(filteredRewards.length ? filteredRewards : rewards)

  // const rewardsToDisplay = orderedRewards.length ? orderedRewards : filteredRewards.length ? filteredRewards : rewards

  // let rewardsToDisplay

  // if (orderedRewards.length) {
  //   rewardsToDisplay = orderedRewards
  // } else if (filteredRewards.length) {
  //   rewardsToDisplay = filteredRewards
  // } else {
  //   rewardsToDisplay = rewards
  // }

  /** END */

  return (
    <>
      <Styled.Toolbar>
        <ToggleGroup 
          id="order-by" 
          values={LIST_MUTATION_KEYS} 
          label="Sort Rewards" 
          onChange={onOrderKeyChange} 
        />
        {/* <ToggleGroup id="filter-by" values={LIST_MUTATION_KEYS} label="Filter Rewards" onChange={onFilterKeyChange} /> */}
      </Styled.Toolbar>
      <Styled.Separator decorative />
      <Styled.List>
        {rewardsToDisplay.map(reward => (
          <li key={reward.id}>
            <RewardCard reward={reward} />
          </li>
        ))}
      </Styled.List>
    </>
  )
}

export default RewardsList