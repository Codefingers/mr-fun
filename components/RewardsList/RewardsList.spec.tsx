import { Providers, render } from '../../util/testing'
import RewardsList from './RewardsList'
import { mockedRewards } from './__fixtures__'

const props = {
  rewards: mockedRewards
}

describe('<RewardsList />', () => {
  it('Should render without having an existential crisis', () => {
    const { container } = render(<RewardsList {...props} />, { wrapper: Providers })

    expect(container).toMatchSnapshot()
  })
})