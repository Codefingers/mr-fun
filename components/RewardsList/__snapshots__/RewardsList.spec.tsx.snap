// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RewardsList /> Should render without having an existential crisis 1`] = `
<div>
  <div
    class="sc-ifAKCX bTNgqR"
  >
    <span
      class="sc-htpNat hdiYuO"
      id="radix-0"
      role="label"
    >
      Sort Rewards
    </span>
    <div
      dir="ltr"
      role="group"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-checked="false"
        class="sc-bwzfXH mLdsH"
        data-radix-collection-item=""
        data-state="off"
        role="radio"
        tabindex="-1"
        type="button"
      >
        quantity
      </button>
      <button
        aria-checked="false"
        class="sc-bwzfXH mLdsH"
        data-radix-collection-item=""
        data-state="off"
        role="radio"
        tabindex="-1"
        type="button"
      >
        points
      </button>
      <button
        aria-checked="false"
        class="sc-bwzfXH mLdsH"
        data-radix-collection-item=""
        data-state="off"
        role="radio"
        tabindex="-1"
        type="button"
      >
        capPerReviewifyMember
      </button>
    </div>
  </div>
  <div
    class="sc-EHOje jcYroV"
    data-orientation="horizontal"
    role="none"
  />
  <ul
    class="sc-bxivhb kkLDcn"
  >
    <li>
      <div
        class="sc-bdVaJa jcXhbk"
      >
        <h2>
          __TEST_REWARD_NAME_1__
        </h2>
        <img
          alt=""
          src="__TEST_IMAGE_URL_1__"
        />
        <p>
          __TEST_DESCRIPTION_1__
        </p>
      </div>
    </li>
    <li>
      <div
        class="sc-bdVaJa jcXhbk"
      >
        <h2>
          __TEST_REWARD_NAME_2__
        </h2>
        <img
          alt=""
          src="__TEST_IMAGE_URL_2__"
        />
        <p>
          __TEST_DESCRIPTION_2__
        </p>
      </div>
    </li>
    <li>
      <div
        class="sc-bdVaJa jcXhbk"
      >
        <h2>
          __TEST_REWARD_NAME_3__
        </h2>
        <img
          alt=""
          src="__TEST_IMAGE_URL_3__"
        />
        <p>
          __TEST_DESCRIPTION_3__
        </p>
      </div>
    </li>
  </ul>
</div>
`;
